### PR TITLE
Improved solo high scores, basic band high scores

### PR DESCRIPTION
**Pull request by thomeval**
This PR improves the way solo high scores are handled on the Music Library screen (and elsewhere), bringing it more in-line with RB and CH.

image

Main changes include:
- Solo high scores now only display if there is exactly one non-bot player connected.
- The Music Library will no longer display irrelevant scores. It will only display the best score for the current player, for their current instrument.
- Removed the now unnecessary instrument icon.
- The existing behaviour of showing both the best score and best percentage, even if those are from different runs, has been left intact.
- Replaced the difficulty display with new icons, bringing us one step closer to the proposed redesign of this screen (thanks @kaduwaengertner ).
- FC's are now shown with a gold percentage.
- The "New High Score" notification will now trigger if the given player has beaten their own personal best, not the "machine best".

This PR also introduces a basic version of band high scores, shown when there are no players, or more than one player connected:

image

These are pulled from the `GameRecord` table. Known limitations:
- Band high scores include scores obtained by only one player. I propose that we only display scores with at least two players, but we don't currently store the number of band members in the database.
- Bots are welcome to set high scores. I propose that we exclude any high score that had any bot players, but we don't currently track which scores had bot involvement in the database. This will be addressed in PR #898 .

### CODE REVIEW

**Comment by thomeval** [comment]
Not yet ready to merge. Still needs testing.

**Comment by thomeval** [comment]
Ready for review.

File: Assets/Script/Menu/MusicLibrary/MusicLibraryMenu.cs [hunk 1]
```diff
@@ -223,6 +224,8 @@ protected override List<ViewType> CreateViewList()
             return viewList;
         }
 
+        private bool _shouldDisplaySoloHighScores => PlayerContainer.Players.Count(e => !e.Profile.IsBot) == 1;
```
Comment (EliteAsian123) [inline Assets/Script/Menu/MusicLibrary/MusicLibraryMenu.cs:227]: Properties should always be in `PascalCase`

Comment (EliteAsian123) [inline Assets/Script/Menu/MusicLibrary/MusicLibraryMenu.cs:227]: Properties should always be in `PascalCase`.

File: Assets/Script/Menu/MusicLibrary/ViewTypes/SongViewType.cs [hunk 2]
```diff
@@ -47,34 +49,42 @@ public override string GetSecondaryText(bool selected)
 #nullable enable
         public override Sprite? GetIcon()
 #nullable disable
-        {
+        {
```
Comment (EliteAsian123) [inline Assets/Script/Menu/MusicLibrary/ViewTypes/SongViewType.cs:52]: Make sure to strip extra whitespace

File: Assets/Script/Scores/ScoreContainer.cs [hunk 3]
```diff
@@ -120,6 +120,51 @@ public static void RecordScore(GameRecord gameRecord, List<PlayerScoreRecord> pl
...
+            {
+                return;
+            }
+
+            BandHighScores[songChecksum] = currentBandScore;
+        }
+
+        private static void UpdatePlayerHighScores(HashWrapper songChecksum, PlayerScoreRecord newScore)
+        {
+
```
Comment (EliteAsian123) [inline Assets/Script/Scores/ScoreContainer.cs:140]: Remove this empty line

File: Assets/Script/Scores/ScoreContainer.cs [hunk 4]
```diff
@@ -120,6 +120,51 @@ public static void RecordScore(GameRecord gameRecord, List<PlayerScoreRecord> pl
...
+                    PlayerHighScoresByPct[songChecksum] = newScore;
+                }
+            }
+
+            // Update the best percentage entry if appropriate.
+            if (newScore.Difficulty == currentBestPct.Difficulty || newScore.GetPercent() > currentBest.GetPercent())
+            {
+                PlayerHighScoresByPct[songChecksum] = newScore;
+            }
+
```
Comment (EliteAsian123) [inline Assets/Script/Scores/ScoreContainer.cs:165]: Remove this empty line

**Review by EliteAsian123 — CHANGES_REQUESTED** [review]
Hi, sorry for the very late review, I've been very busy recently.

I've listed some styling changes below, however, there's one big thing that I want changed which is how bots are handled at the moment with band scoring. 

Bots should *never* under any circumstances have their scores saved. This wasn't really a problem before, as we only displayed individual scores, but with this change, bands with bot scores in there show up, which obviously doesn't make sense. Once we introduce band multipliers, it will make even less sense, as individual scores will be effected too.

Because of this, all band scores with bots should be completely invalidated and not saved. From the looks of it we currently do, however, I'd classify this as a bug as it really shouldn't be. 

Once this change is made, this PR looks pretty much good to me, however I haven't tested it in-game yet so I'll do that once this is fixed.

Thanks!

Edit: if this is changed then also the `PlayerContainer.Players.Count(e => !e.Profile.IsBot) == 1` wouldn't be needed and could be simplified as if a bot is in your band, you're still technically in a band and not playing solo.

**Comment by EliteAsian123** [comment]
#898 has been merged, so if you can update this PR to account for this, I'll review it and merge it. Thanks!

**Comment by thomeval** [comment]
@EliteAsian123 Okay, this PR has been updated with the latest dev branch and is ready to review.

**Pull request merged**